Feature/fv/error handling

favillat opened the pull request.

ETSells committed:
Merge pull request #149 from SunDevilRocketry/feature/BB/fin-calib-motor-snap

Feature/bb/fin calib motor snap

ETSells committed:
Revert "Feature/bb/fin calib motor snap"

ETSells committed:
Merge pull request #153 from SunDevilRocketry/revert-149-feature/BB/fin-calib-motor-snap

Revert "Feature/bb/fin calib motor snap"

favillat committed:
Added fail safes to avoid fail fast errors
in flight descent, in_flight, launch_detect checked flash_status, if its anything other than FLASH_OK, break and turn red

favillat committed:
Merge remote-tracking branch 'origin' into feature/Fv/error-handling

favillat committed:
Revert "Added fail safes to avoid fail fast errors"

This reverts commit 1a8a5246cfd0eded6424b23f7a04e408dfd69093.

favillat committed:
reverted prelaunch to dev/appa branch

favillat committed:
Revert "Added fail safes to avoid fail fast errors"

This reverts commit 1a8a5246cfd0eded6424b23f7a04e408dfd69093.

favillat committed:
Added error handling in flight.c

favillat committed:
Merge branch 'development/appa' of https://github.com/SunDevilRocketry/Flight-Computer-Firmware into feature/Fv/error-handling

favillat committed:
updated mod submodule

favillat committed:
failfast error prevention
-on a flash fail disble future flashing and set led to red
-hopefully doesnt mess with timing as it is kinda gaurded by flash_logging_enabled bool

favillat committed:
submodule changes

favillat committed:
asdasd

ETSells reviewed (changes requested):
This pull request pulls up the `mod` submodule. If there were changes made to mod under this ticket, please make sure there is a pull request from the submodule, otherwise please discard the submodule changes.

favillat committed:
fixed mod submodule changes

favillat committed:
Merge branch 'development/appa' of https://github.com/SunDevilRocketry/Flight-Computer-Firmware into feature/FV/error-handling

ETSells reviewed (changes requested):
Just a few requests here. I think this is a good way of approaching it, it's just worth addressing the testability part and making sure we don't have any hardcoded config.

favillat committed:
Addressed first round of comments left on pr
-removed flash_logging_enabled as it was uneeded
-writes and erases directly update *flash_status
-before any flash ops we now check for FLASH_OK (on fail writes and erases are skipped)
-LED changed from RED to BLUE tp indicate soft logging error/full flash

ETSells commented on app/flight/appa/rev2/flight.c:
```diff
@@ -145,34 +145,64 @@ if ( *sensor_status != SENSOR_OK )
+            {
+            if (*flash_status != FLASH_OK)
+                {   
+                /* stop future logs */
```
Change requested: This branch will only get hit if flash is busy. Be careful here

ETSells commented on app/flight/appa/rev2/flight.c:
```diff
@@ -330,15 +370,29 @@ if ( *sensor_status != SENSOR_OK )
+            
+            *flash_status = store_frame( flash_handle, &sensor_data, current_timestamp, flash_address );
+                
+            if(*flash_status != FLASH_OK)
```
again since you check FLASH_OK above, this branch is unlikely to be hit. you can reduce complexity by moving this logic.

ETSells reviewed (changes requested):
Main change request is for bad braces (inside the "while flash busy" block).
Also some recommendations for reducing complexity.

favillat committed:
Reduced Complexity and fixed braces/whitespaces

ETSells commented on app/flight/appa/rev2/flight.c:
```diff
@@ -223,15 +241,25 @@ if ( apogee_detect() )
-    /* Write to flash */
-    while( flash_is_flash_busy() == FLASH_BUSY )
+
+    if(*sensor_status == SENSOR_OK )
```
Change requested: we do a sensor status check earlier, so the FALSE branch here will be unreachable.

ETSells commented on app/flight/appa/rev2/flight.c:
```diff
@@ -223,15 +241,25 @@ if ( apogee_detect() )
+            {
+                *flash_status = store_frame( flash_handle, &sensor_data, current_timestamp, flash_address );
+                  
+                if(*flash_status != FLASH_OK)
```
Change Requested: Consider performing the check before you store frame (on 241). Since you don't check it before you store, you can end up reaching the store_frame call above which is the behavior we're trying to stop. That way, you can remove the branching here as well and set the last_flash_timestamp outside of additional branching behavior.

ETSells commented on app/flight/appa/rev2/flight.c:
```diff
@@ -330,15 +358,22 @@ if ( *sensor_status != SENSOR_OK )
-    /* Write to flash */
-    while( flash_is_flash_busy() == FLASH_BUSY )
+
+    if(*flash_status == FLASH_OK)
```
Consider pushing this check to the `if ( flash_handle->address...` above (&& the two conditions).

The reason for this is to reduce cyclomatic complexity (every new if statement is another branch. We don't need condition-decision coverage, we just need line coverage) as well as to make sure the color is set correctly (think about how execution will work in an error case. The next time you get to this function, it'll set the color purple so it never goes blue).

ETSells reviewed (changes requested):
I'm so sorry I keep making you do revisions on this. Might be good to pair-program it at one of my office hours next week so I don't keep springing more considerations on you?

Basically the requested changes boil down to the question: "What happens in every possible branch here?" (For the record: this is something I try to do every time I modify code in our codebase, which is one of the reasons why it's so important that we keep CC low!)

There are a couple of cases here where there'll be unintentional behavior (usually related to status LED color) in like one or two possible paths of execution. I've done my best to lay  all these out but I might be missing some.

favillat commented:
it's totally fine no worries, i apologize for having you review this so many times as well lol, i'll work on the changes you mentioned but i'll definitely drop into the shop on tuesday to go over it with you before i make any more pr!

favillat committed:
fixed minor details - still reducing CC

favillat committed:
checkpoint

favillat committed:
reducd CC branching (hopefully)
Descent and inflight now check flash status inline with flash full
removed unreachable sensor status check from in flight
removed busy wait from launch detect timeout
fixed some more whitespace inconsistencies

favillat committed:
whitespaces and comments

ETSells commented on app/flight/appa/rev2/flight.c:
```diff
@@ -145,34 +144,51 @@ if ( *sensor_status != SENSOR_OK )
+
 
-    /* Reset the timer      */
-    *launch_detect_start_time = HAL_GetTick();
```
This might still be necessary

ETSells reviewed (approved):
I think this looks right. Recommend testing on the FC and verifying everything still works as expected before you merge.

Great job Fernando!
